[NEMO-253] Refactor getInternal(Main/Additional)OutputMap in TaskExecutor

apeinot opened the pull request:
JIRA: [NEMO-253: Refactor getInternal(Main/Additional)OutputMap in TaskExecutor](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-253)

**Major changes:**
- Merge methods getInternalMainOutputs and getInternalAdditionalOutputMap from TaskExecutor into one method called getInternalOutputMap

**Minor changes to note:**
- Small change in TaskExecutor::prepare to parse the new method output

**Tests for the changes:**
- All the existing tests still pass
- Add one test which create a TaskExecutor with null inputs to test the exception

Closes #196

johnyangk commented:
Hi @apeinot! Thanks for submitting this PR and welcome to the community!

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/task/TaskExecutor.java:
```diff
@@ -213,13 +213,18 @@ public TaskExecutor(final Task task,
-      final List<NextIntraTaskOperatorInfo> internalMainOutputs =
-        getInternalMainOutputs(irVertex, irVertexDag, edgeIndexMap, operatorWatermarkManagerMap);
+      final List<NextIntraTaskOperatorInfo> internalMainOutputs;
+      final String key = "main";
```
How about moving this into `AdditionalOutputTagProperty` as a public static variable, to ensure there's only a single value for the main output tag? This also has the performance benefit of caching the result of hashCode() and reusing it across all task executions in the executor.

Can you also add a TODO in the comment for the JIRA I've just created? It has become more clear to me after seeing this PR that it may be worthwhile to make the "main" tag more explicit.
https://issues.apache.org/jira/browse/NEMO-348

So I'm thinking something like the following. Let me know if you think this makes sense!
```java
// TODO #348: Consider OutputTagProperty
public static String MAIN_OUTPUT_TAG = "main"
```

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/task/TaskExecutor.java:
```diff
@@ -540,20 +545,34 @@ private boolean handleDataFetchers(final List<DataFetcher> fetchers) {
+          if (isPresent) {
+            outputTag = edge.getPropertyValue(AdditionalOutputTagProperty.class).get();
+          } else {
+            outputTag = "main";
```
Maybe refer to `AdditionalOutputTagProperty.MAIN_OUTPUT_TAG`?
(See comment above)

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/task/TaskExecutorTest.java:
```diff
@@ -188,6 +189,24 @@ public void close() throws IOException {
+      final TaskExecutor taskExecutor = getTaskExecutor(null, null);
+      taskExecutor.execute();
+
+      // Check the output.
```
Maybe change to:
// This should not be reached

Since we expect `taskExecutor.execute()` to throw `NullPointerException`?

apeinot commented:
Hi @johnyangk ! Thanks for your advice. I think your proposed change is relevant and I have addressed it in my last commit. 
There is only one difference. Because of the checkstyle configuration, I had to change MAIN_OUTPUT_TAG to mainOutputTag and made this variable private instead of public (I added the corresponding getter)

johnyangk commented:
Thanks! I've merged it.
